src/main/java/net/snowflake/client/jdbc/SnowflakeBaseResultSet.java

```diff
@@ -1347,11 +1361,83 @@ public <T> T getObject(int columnIndex, Class<T> type) throws SQLException {
       SQLInput sqlInput = (SQLInput) getObject(columnIndex);
       instance.readSQL(sqlInput, null);
       return (T) instance;
+    } else if (Map.class.isAssignableFrom(type)) {
+      JsonNode jsonNode = ((JsonSqlInput) getObject(columnIndex)).getInput();
+      return (T) OBJECT_MAPPER.convertValue(jsonNode, new TypeReference<Map<String, Object>>() {});
     } else {
       return (T) getObject(columnIndex);
     }
   }
 
+  public <T> List<T> getList(int columnIndex, Class<T> type) throws SQLException {
+    T[] sqlInputs = getArray(columnIndex, type);
+    return Arrays.asList(sqlInputs);
+  }
+
+  public <T> T[] getArray(int columnIndex, Class<T> type) throws SQLException {
+    Optional<Supplier<SQLData>> typeFactory = SnowflakeObjectTypeFactories.get(type);
+    Map[] sqlInputs = (Map[]) getArray(columnIndex).getArray();
+    T[] arr = (T[]) java.lang.reflect.Array.newInstance(type, sqlInputs.length);
+    AtomicInteger counter = new AtomicInteger(0);
```

Comment:
What do you think about reversing the flow - instead of creating list from array - create list first and then array? I believe this may make the code easier (i.e. getting rid off this counter).

Reply:
It was refactored. For me method getArray is base because it exists in jdbc api. Please take a look now because without streams it doesn't look like that writing getList first could be more readable